File: Python/BFS.py
from collections import deque

graph = {}

graph['a'] = ['b', 'c']
graph['b'] = ['a', 'c', 'd']
graph['c'] = ['a']
graph['d'] = ['b', 'e']
graph['e'] = ['d']

# Disconnected bit:
graph['f'] = ['z']
graph['z'] = ['f']

# Visited to mark the visited nodes and queue to achieve BFS traversal
visited = set()
dist = {}
parent = {}

def bfs_start(graph, s):
    q = deque()
    
    # Setting up my start node
    visited.add(s)
    q.append(s)
    dist[s] = 0  # Also resembles the tree level
    parent[s] = None
    
    while q:
        current = q.popleft()
        for neighbor in graph[current]:
            if neighbor not in visited:
                print(neighbor)
                # Visit that neighbor, and add it to the queue to visit its neighbors in the next level
                visited.add(neighbor)
                q.append(neighbor)
                
                # The neighbor's level is that of its parent + 1, and its parent is the one that just got popped from queue
                dist[neighbor] = dist[current] + 1
                parent[neighbor] = current
                
            
def bfs(graph):
    for node in graph.keys():
        if node not in visited:
            bfs_start(graph, node)


bfs(graph)

print("Distances:", dist)
print("Parents:", parent)

File: CPP/BFS.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <queue>
#include <unordered_set>

using namespace std;

// Define the graph using an adjacency list
class Graph
{
public:
    unordered_map<int, vector<int>> adjList;

    // Add an edge to the graph
    void addEdge(int u, int v, bool isDirected = false)
    {
        adjList[u].push_back(v); // Add an edge from u to v
        if (!isDirected)
        {
            adjList[v].push_back(u); // For undirected graphs, add v to u's list
        }
    }

    // Perform BFS
    void bfs(int start)
    {
        queue<int> q;               // Queue for BFS
        unordered_set<int> visited; // Track visited nodes

        // Start BFS from the given node
        q.push(start);
        visited.insert(start);

        cout << "BFS Traversal: ";
        while (!q.empty())
        {
            int current = q.front();
            q.pop();
            cout << current << " ";

            // Add all unvisited neighbors to the queue
            for (int neighbor : adjList[current])
            {
                if (visited.find(neighbor) == visited.end())
                {
                    q.push(neighbor);
                    visited.insert(neighbor);
                }
            }
        }
        cout << endl;
    }

    // Print the adjacency list
    void printGraph()
    {
        for (auto &[node, neighbors] : adjList)
        {
            cout << node << ": ";
            for (auto &neighbor : neighbors)
            {
                cout << neighbor << " ";
            }
            cout << endl;
        }
    }
};

int main()
{
    Graph g;

    // Add edges to the graph (example graph)
    g.addEdge(1, 2);
    g.addEdge(1, 3);
    g.addEdge(2, 4);
    g.addEdge(2, 5);
    g.addEdge(3, 6);
    g.addEdge(3, 7);

    // Print the adjacency list
    g.printGraph();

    // Perform BFS starting from node 1
    g.bfs(1);

    return 0;
}
